### ISSUE TITLE
None of the common-lisp distros working  

### ISSUE THREAD

**Issue by nilqed** [Microsoft/WSL#1128]
Please use the following bug reporting template to help produce actionable and reproducible issues. Please try to ensure that the reproduction is minimal so that the team can go through more bugs!
- A brief description
  I've tried to install/compile the following well known "common-lisp" distributions: SBCL, ECL, CCL, CLISP.
  I didn't manage to get one of them working. 
- Expected results
  To see a prompt, obviously ;)
- Actual results (with terminal output if applicable)

kfp@AURORA:~$ ecl
ecl: error while loading shared libraries: libecl.so.16.1: cannot enable executable stack as shared object requires: Invalid argument

kfp@AURORA:~$ clisp
/usr/lib/clisp-2.49/base/lisp.run: error while loading shared libraries: libavcall.so.0: cannot enable executable stack as shared object requires: Invalid argument

kfp@AURORA:~$ sbcl
WARNING:
Couldn't re-execute SBCL with proper personality flags (/proc isn't mounted? setuid?)
Trying to continue anyway.
This is SBCL 1.3.0, an implementation of ANSI Common Lisp.
More information about SBCL is available at http://www.sbcl.org/.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
(hangs) ^Z
[1]+  Angehalten              sbcl

kfp@AURORA:~/ccl$ ./lx86cl64
kfp@AURORA:~/ccl$ #### terminates after about  10s without output

SBCL and CLISP may be installed with apt or manually from the sites below:

SBCL: http://www.sbcl.org/
ECL: https://common-lisp.net/project/ecl/static/files/
CCL: http://ccl.clozure.com/
CLISP: http://www.gnu.org/software/clisp/
- Your Windows build number
  14393.105
- Steps / All commands required to reproduce the error from a brand new installation 
  e.g. sudo apt install clisp 
  or sudo apt install sbcl
- Strace of the failing command
- Required packages and commands to install
  ggc, autoconf, make, [svn]

See [our contributing instructions](/CONTRIBUTING.md) for assistance.

BTW: Great thing :)

**Comment by bvnierop** [comment]
I just ran into this as well.

**Comment by mobluse** [comment]
Same problem with in maxima, xmaxima, and wxmaxima that I believe use Common Lisp.
```
$ maxima
personality change failure 0
```
I use Build 14931.1000.

**Comment by BRAVEMAN-L-BRID** [comment]
same problem 
$ clisp
/usr/lib/clisp-2.49/base/lisp.run: error while loading shared libraries: libavcall.so.0: cannot enable executable stack as shared object requires: Invalid argument

**Comment by pdiazs** [comment]
I have the same problem . Just to say that this invalidates the use of maxima and wxmaxima  and represents a limitation in using Texmacs

**Comment by showtimesynergy** [comment]
I can second this, sbcl complains about the existence of /proc/sys/binfmt_misc

**Comment by benhillis** [comment]
@showtimesynergy - Can you give me some repro steps I can follow to see this if this still reproduces?  I have a binfmt_misc related fix I want to validate.

**Comment by showtimesynergy** [comment]
@benhillis The only time it mentions it is during install via `sudo apt-get install sbcl`

~~~~
$ sudo apt-get install sbcl
Reading package lists... Done
Building dependency tree
Reading state information... Done
Suggested packages:
  sbcl-doc sbcl-source slime
The following NEW packages will be installed:
  sbcl
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/6,757 kB of archives.
After this operation, 51.4 MB of additional disk space will be used.
Selecting previously unselected package sbcl.
(Reading database ... 49002 files and directories currently installed.)
Preparing to unpack .../sbcl_2%3a1.1.14-2_amd64.deb ...
Unpacking sbcl (2:1.1.14-2) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Setting up sbcl (2:1.1.14-2) ...
update-binfmts: warning: binfmt_misc module seemed to be loaded, but no /proc/sys/fs/binfmt_misc directory!  Giving up.
~~~~
attempting to run sbcl after this displays the following and hangs
~~~~
$ sbcl
WARNING:
Couldn't re-execute SBCL with proper personality flags (/proc isn't mounted? setuid?)
Trying to continue anyway.
This is SBCL 1.1.14.debian, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
~~~~

**Comment by benhillis** [comment]
@showtimesynergy - I'm no longer seeing the error about binfmt and I'm seeing the sbcl entry under /proc/sys/fs/binfmt_misc.

However when running sbcl it looks like the issue has to do with a call to personality(ADDR_NO_RANDOMIZE) which disabling randomizing of the address space.

**Comment by stehufntdev** [comment]
Support for ADDR_NO_RANDOMIZE should be out in an insider build soon.  Locally I confirmed that it resolved the sbcl warning but it still hung so we'll need to investigate further.  Please give us feedback on the user voice page so we can prioritize common lisp - https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/category/161892-bash

**Comment by mobluse** [comment]
There was no Lisp-related topic that you could vote on so I added one: https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/17629492-maxima-and-common-lisp-should-work

**Comment by oldk1331** [comment]
Do not install SBCL from Ubuntu repo, they are too old.  Recent version of
SBCL (1.3.12, 1.3.15) works on WSL, but can not dump images.

See:
https://sourceforge.net/p/sbcl/mailman/message/35708531/
https://sourceforge.net/p/sbcl/mailman/message/35708604/

**Comment by Shadowys** [comment]
MKCL works.

**Comment by bhaible** [comment]
Recent versions of clisp (from its hg repository) and of libffcall (from https://alpha.gnu.org/gnu/libffcall/) should no longer produce the "cannot enable executable stack as shared object requires" error.

**Comment by sngeth** [comment]
Still see this error when starting clisp using (1:2.49-9ubuntu1) latest package provided by Ubuntu 16.04. 

Tried to install 2.49 from source and i had to configure the makefile without libsigsegv as it errorered about that. After that I ran make and it blew up with this :-(
`*** - UNIX error 5 (EIO): I/O errorMakefile:1397: recipe for target 'interpreted.mem' failed`

**Comment by bhaible** [comment]
_> Tried to install 2.49 from source_
Try a recent version of clisp instead. Maybe this one works: https://launchpad.net/ubuntu/artful/amd64/clisp/1:2.49.60+-2
For a recent version of libffcall, use libffcall-1.13 (from https://ftp.gnu.org/gnu/libffcall/).

**Comment by jstarks** [comment]
In internal builds, I have fixed the executable stack and EIO failures with clisp, and the hang with sbcl. These were three separate issues, surprisingly.

These fixes should be available in a future insiders build.

**Comment by mobluse** [comment]
Does this fix also make it possible to run Maxima?
`sudo apt install maxima`
Now one gets error:
```
$ maxima
personality failure 38
```

Microsoft Windows [Version 10.0.17025.1000]

**Comment by jstarks** [comment]
Nope, this is a different problem. I'll see if that's an easy fix.

**Comment by jstarks** [comment]
This should be fixed in [insider build 17063](https://docs.microsoft.com/en-us/windows/wsl/release-notes).

**Comment by terminalcommand** [comment]
SBCL still does not work, it doesn't show the /proc not mounted error anymore. But the line starting with a * (star) does not come, I can type statements e.g. (+ 3 2) but they are not evaluated. If I hit C-C it goes into the debugger and gets  trapped in a debug loop. 

However in the interim, I've found a solution. I installed the windows version of SBCL and then made an alias like this: 

`alias sbcl=sbcl.exe`

Now I can use sbcl inside WSL, albeit as a windows executable.

**Comment by GabrielDosReis** [comment]
I confirmed that I build CLISP and ECL successfully, and used those lisps to build [OpenAxiom](http://www.open-axiom.org).

However, SBCL remains broken.  There are new analysis in [this thread](https://sourceforge.net/p/sbcl/mailman/sbcl-devel/thread/CAAsgrApF2aPiubm9bgh-ZNTDg_PTdUMFHOUuEDqX36b8fQSA%2BQ%40mail.gmail.com/#msg36221859)

It sounds like WSL might have a weak support for SIGSEGV?

**Comment by oldk1331** [comment]
As I reported in Mar 8, 2017, SBCL (1.3.12, 1.3.15) can run in WSL, but can't dump image (`save-lisp-and-die`), thus can't be built. According to jstarks, this bug seems to be fixed in next version? Has anyone tried in Windows Insider?  Anyway, I'll find out when Windows 1803 get released next month.

Just FYI, there seems to be a regression in recent SBCL versions: starting with SBCL 1.3.21, the prebuilt windows binaries can't run in WSL, it hangs.

**Comment by GabrielDosReis** [comment]
I am running [17093](https://blogs.windows.com/windowsexperience/2018/02/07/announcing-windows-10-insider-preview-build-17093-pc/).

The previous fixes allow building ECL and CLISP.  I don't think SBCL was fixed.  See the thread on the SBCL list I referenced earlier, it has useful analysis.

**Comment by bhaible** [comment]
_> It sounds like WSL might have a weak support for SIGSEGV?_

I can't confirm this. CLISP relies on SIGSEGVs as well for its GC (via libsigsegv), and in the [current release branch](https://sourceforge.net/p/clisp/clisp/ci/clisp-2.50/tree/) it builds fine and passes its test suite on WSL (Windows build 16299). That is, it catches many many SIGSEGVs and proceeds without problems, just as on real Linux.

**Comment by therealkenc** [comment]
If SBCL is still problematic, it would be helpful to open an new issue following the template, especially the repro steps from clean Ubuntu install part. This one was closed because "executable stack and EIO failures" were [addressed](https://github.com/Microsoft/WSL/issues/1128#issuecomment-345846553) (plus `ADDR_NO_RANDOMIZE` circa [late 2016](https://github.com/Microsoft/WSL/issues/1128#issuecomment-266534906)). 

There _might_ be problems with SIGSEGV handling per #2555 ([message](https://github.com/Microsoft/WSL/issues/2555#issuecomment-336510125)) et al. But bear in mind the "pretty sure" caveat in that message over there, and the timeframe of the post. I don't use SBCL and I haven't confirmed whether it is relevant. Sounds from the post above that `libsigsegv` could be just fine. I only know mcontext_t.uc_mcontext doesn't make it into the handler on WSL last time I checked (which isn't recently).
